### PR TITLE
[pytorch] Add roll support

### LINKED ISSUE #3734: Support roll operation for ndarray
## Description
Since torch or ort ndarrays can perform roll operations on its ndarrays, I think is a good enhancement to add that operation to ndarray api. 

I know is not hard to implement a custom function, but a default implementation will help the community a lot.

### PR DESCRIPTION
**Pull request by serhiy-bzhezytskyy**
## Description ##

Adds `NDArray.roll(shifts, axes)`, matching numpy's `np.roll`: elements are shifted along the given axes, with elements shifted past the last position reintroduced at the start. With no axes, this `NDArray` is flattened before rolling, then reshaped back to its original shape. `roll(shift)` and `roll(shift, axis)` are convenience defaults delegating to the one abstract primitive, so only that needs a per-engine implementation -- mirroring the `split` overloads already in `NDArray`.

Only the pytorch engine implements it (via `torch::roll`); mxnet, tensorflow and the rust extension throw `UnsupportedOperationException`, matching the existing pattern for engine-specific ops like `diff` (#3663).

Closes #3734

### Verification

Rebuilt the pytorch JNI native library from source (`./gradlew :engines:pytorch:pytorch-native:compileJNI`) and ran the new integration test (`NDArrayNumericOpTest.testRoll`) through the freshly compiled library, not a cached one -- covering flatten roll, negative shift, and axis roll. Full `NDArrayNumericOpTest` class still passes (29/29, 3 pre-existing unsupported cases unrelated to this change).

### CODE REVIEW

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/jni/JniUtils.java [hunk 1]
```diff
@@ -1945,4 +1945,11 @@ public static PtNDArray diff(PtNDArray ndArray, int n, int dim) {
         return new PtNDArray(
                 ndArray.getManager(), PyTorchLibrary.LIB.torchDiff(ndArray.getHandle(), n, dim));
     }
+
+    public static PtNDArray roll(PtNDArray ndArray, long[] shifts, int[] axes) {
+        long[] dims = Arrays.stream(axes).mapToLong(i -> i).toArray();
+        return new PtNDArray(
+                ndArray.getManager(),
+                PyTorchLibrary.LIB.torchRoll(ndArray.getHandle(), shifts, dims));
```
[the remark this answers is not in the document]
Comment (serhiy-bzhezytskyy) [reply engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/jni/JniUtils.java:1953]: Confirmed empirically before changing anything: rolled a reshaped (2, 4) array with no axis through the actual compiled library -- input [[0,1,2,3],[4,5,6,7]], roll(2) gives [[6,7,0,1],[2,3,4,5]], correct flatten+roll+reshape. torch::roll special-cases an empty dims to mean "flatten", so shifts.length()==1 with dims.length()==0 is valid there, not a length mismatch. Added that exact case as a permanent assertion in the integration test (see the reply below) instead of changing the native call.

File: integration/src/main/java/ai/djl/integration/tests/ndarray/NDArrayNumericOpTest.java [hunk 2]
```diff
@@ -788,4 +788,24 @@ public void testDiff() {
...
+            // No axis: this NDArray is flattened before rolling, then reshaped back
+            NDArray expected = manager.create(new float[] {6f, 7f, 0f, 1f, 2f, 3f, 4f, 5f});
+            Assert.assertEquals(expected, array.roll(2));
+
+            // Negative shift rolls in the opposite direction
+            expected = manager.create(new float[] {2f, 3f, 4f, 5f, 6f, 7f, 0f, 1f});
+            Assert.assertEquals(expected, array.roll(-2));
+
+            // Along a given axis
+            NDArray array2d = array.reshape(2, 4);
```
[the remark this answers is not in the document]
Comment (serhiy-bzhezytskyy) [reply integration/src/main/java/ai/djl/integration/tests/ndarray/NDArrayNumericOpTest.java]: Confirmed, fixed: added a case rolling a reshaped (2, 4) array with no axis, verified against the same libtorch build ([[0,1,2,3],[4,5,6,7]] -> [[6,7,0,1],[2,3,4,5]]).

File: api/src/main/java/ai/djl/ndarray/NDArrayAdapter.java [hunk 3]
```diff
@@ -1308,6 +1308,12 @@ public NDArray diff(int n, int dim) {
         throw new UnsupportedOperationException("Not implemented yet.");
     }
 
+    /** {@inheritDoc} */
+    @Override
+    public NDArray roll(long[] shifts, int[] axes) {
+        throw new UnsupportedOperationException("Not implemented yet.");
```
[the remark this answers is not in the document]
Comment (serhiy-bzhezytskyy) [reply api/src/main/java/ai/djl/ndarray/NDArrayAdapter.java:1314]: "Not implemented yet." is the message every one of the other 18 stub overrides in NDArrayAdapter already throws verbatim (diff, conj, etc.) -- singling out roll would be inconsistent with the rest of the file rather than an improvement. Leaving as-is to match the existing pattern.

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/jni/JniUtils.java [hunk 4]
```diff
@@ -1945,4 +1945,11 @@ public static PtNDArray diff(PtNDArray ndArray, int n, int dim) {
         return new PtNDArray(
                 ndArray.getManager(), PyTorchLibrary.LIB.torchDiff(ndArray.getHandle(), n, dim));
     }
+
+    public static PtNDArray roll(PtNDArray ndArray, long[] shifts, int[] axes) {
+        long[] dims = Arrays.stream(axes).mapToLong(i -> i).toArray();
```
[the remark this answers is not in the document]
Comment (serhiy-bzhezytskyy) [reply engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/jni/JniUtils.java:1950]: This is the same int[]-to-long[] conversion already used twice elsewhere in this file (lines 911 and 1930) -- following the existing pattern here rather than introducing a third, different idiom.

**Comment by xyang16** [comment]
@serhiy-bzhezytskyy Thanks for contribution! Could you please run `./gradlew formatJava` to format code?

**Comment by serhiy-bzhezytskyy** [comment]
Thanks, fixed.

File: engines/pytorch/pytorch-native/src/main/native/ai_djl_pytorch_jni_PyTorchLibrary_torch_isjm.cc [hunk 5]
```diff
@@ -193,3 +193,14 @@ JNIEXPORT jlongArray JNICALL Java_ai_djl_pytorch_jni_PyTorchLibrary_torchUnique(
...
   API_END_RETURN()
 }
+
+JNIEXPORT jlong JNICALL Java_ai_djl_pytorch_jni_PyTorchLibrary_torchRoll(
+    JNIEnv* env, jobject jthis, jlong jhandle, jlongArray jshifts, jlongArray jdims) {
+  API_BEGIN()
+  const auto* tensor_ptr = reinterpret_cast<torch::Tensor*>(jhandle);
+  const std::vector<int64_t> shifts = djl::utils::jni::GetVecFromJLongArray(env, jshifts);
+  const std::vector<int64_t> dims = djl::utils::jni::GetVecFromJLongArray(env, jdims);
+  const auto* result_ptr = new torch::Tensor(torch::roll(*tensor_ptr, shifts, dims));
```
Comment (xyang16) [inline engines/pytorch/pytorch-native/src/main/native/ai_djl_pytorch_jni_PyTorchLibrary_torch_isjm.cc]:
```suggestion
  const torch::Tensor* result_ptr = new torch::Tensor(tensor_ptr->roll(shifts, dims));
```

Comment (xyang16) [inline engines/pytorch/pytorch-native/src/main/native/ai_djl_pytorch_jni_PyTorchLibrary_torch_isjm.cc]:
```suggestion
  const torch::Tensor* result_ptr = new torch::Tensor(tensor_ptr->roll(shifts, dims));
```

Comment (xyang16) [reply engines/pytorch/pytorch-native/src/main/native/ai_djl_pytorch_jni_PyTorchLibrary_torch_isjm.cc]: Overall lgtm, just one minor semantic update. Thanks.

Comment (serhiy-bzhezytskyy) [reply engines/pytorch/pytorch-native/src/main/native/ai_djl_pytorch_jni_PyTorchLibrary_torch_isjm.cc]: Applied. Rebuilt the native lib and reran testRoll through it -- passes.

**Review by xyang16 — APPROVED** [review]
